upgrade_analysis/models/upgrade_analysis.py

```diff
@@ -0,0 +1,227 @@
+# Copyright 2011-2015 Therp BV <https://therp.nl>
+# Copyright 2016 Opener B.V. <https://opener.am>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+# flake8: noqa: C901
+
+import logging
+import os
+
+from odoo import fields, models
+from odoo.modules import get_module_path
+from odoo.tools import config
+
+from .. import compare
+
+_logger = logging.getLogger(__name__)
+_IGNORE_MODULES = ["openupgrade_records", "upgrade_analysis"]
+
+
+class UpgradeAnalysis(models.Model):
+    _name = "upgrade.analysis"
+    _description = "Upgrade Analyses"
+
+    analysis_date = fields.Datetime(readonly=True)
+
+    state = fields.Selection(
+        [("draft", "draft"), ("done", "Done")], readonly=True, default="draft"
+    )
+    config_id = fields.Many2one(
+        string="Comparison Config",
+        comodel_name="upgrade.comparison.config",
+        readonly=True,
+        required=True,
+    )
+
+    log = fields.Text(readonly=True)
+    upgrade_path = fields.Char(
+        default=config.get("upgrade_path", False),
+        help="The base file path to save the analyse files of Odoo modules",
+    )
+
+    write_files = fields.Boolean(
+        help="Write analysis files to the module directories", default=True
+    )
+
+    def _get_remote_model(self, connection, model):
+        self.ensure_one()
+        if model == "record":
+            if float(self.config_id.version) < 14:
+                return connection.env["openupgrade.record"]
+            else:
+                return connection.env["upgrade.record"]
+        return False
+
+    def _write_file(
+        self, module_name, version, content, filename="upgrade_analysis.txt"
+    ):
+        module = self.env["ir.module.module"].search([("name", "=", module_name)])[0]
+        if module.is_odoo_module:
+            if not self.upgrade_path:
+                return (
+                    "ERROR: no upgrade_path set when writing analysis of %s\n"
+                    % module_name
+                )
+            module_path = os.path.join(self.upgrade_path, module_name)
+        else:
+            module_path = get_module_path(module_name)
+        if not module_path:
+            return "ERROR: could not find module path of '%s':\n" % (module_name)
+        full_path = os.path.join(module_path, "migrations", version)
+        if not os.path.exists(full_path):
+            try:
+                os.makedirs(full_path)
+            except os.error:
+                return "ERROR: could not create migrations directory %s:\n" % (
+                    full_path
+                )
+        logfile = os.path.join(full_path, filename)
+        try:
+            f = open(logfile, "w")
+        except Exception:
+            return "ERROR: could not open file %s for writing:\n" % logfile
```

Comment:
In general I always make the `odoo` user unable to write into odoo and module source files. I consider that a good security measure, because otherwise a poisoned module could infect all other modules easily.

I think you should be able to configure a specific path to output all this stuff, and that path should be inside odoo's data dir by default. If you really want a module to alter code of other modules, that should be stated explicitly IMHO.

Reply:
Hi @Yajo : this module is not intended to be used in any production instance and should never be used. (see all the warning, when using this module).

It is just made to make analysis, one time, when a new release is out.
The process to write in the odoo folders is the same as for the previous OpenUpgrade version.